File: Java/src/me/metatom/java/string/NFA.java
package me.metatom.java.string;

import edu.princeton.cs.algs4.*;

/**
 * 正则表达式的模式匹配
 *
 * @author igaozp
 * @version 1.0
 * @since 2017-07-28
 */
public class NFA {
    /**
     * 匹配转换
     */
    private char[] re;
    /**
     * epsilon 转换
     */
    private Digraph G;
    /**
     * 状态数量
     */
    private int M;

    /**
     * 构造方法，构建正则表达式的 NFA
     *
     * @param regexp 正则表达式
     */
    public NFA(String regexp) {
        Stack<Integer> ops = new Stack<>();
        re = regexp.toCharArray();
        M = re.length;
        G = new Digraph(M + 1);

        for (int i = 0; i < M; i++) {
            int lp = i;
            if (re[i] == '(' || re[i] == '|') {
                ops.push(i);
            } else if (re[i] == ')') {
                int or = ops.pop();
                if (re[or] == '|') {
                    lp = ops.pop();
                    G.addEdge(lp, or + 1);
                    G.addEdge(or, i);
                } else {
                    lp = or;
                }
            }

            if (i < M - 1 && re[i + 1] == '*') {
                G.addEdge(lp, i + 1);
                G.addEdge(i + 1, lp);
            }

            if (re[i] == '(' || re[i] == '*' || re[i] == ')') {
                G.addEdge(i, i + 1);
            }
        }
    }

    /**
     * NFA 能否检查文本
     *
     * @param txt 检查的文本
     * @return {@code true} 能够检查
     * {@code false} 不能检查
     */
    public boolean recognizes(String txt) {
        Bag<Integer> pc = new Bag<>();
        DirectedDFS dfs = new DirectedDFS(G, 0);

        for (int v = 0; v < G.V(); v++) {
            if (dfs.marked(v)) {
                pc.add(v);
            }
        }

        for (int i = 0; i < txt.length(); i++) {
            Bag<Integer> match = new Bag<>();
            for (int v : pc) {
                if (v < M) {
                    if (re[v] == txt.charAt(i) || re[v] == '.') {
                        match.add(v + 1);
                    }
                }
            }

            pc = new Bag<>();
            dfs = new DirectedDFS(G, match);

            for (int v = 0; v < G.V(); v++) {
                if (dfs.marked(v)) {
                    pc.add(v);
                }
            }
        }

        for (int v : pc) {
            if (v == M) {
                return true;
            }
        }

        return false;
    }

    /**
     * 单元测试
     *
     * @param args 命令行参数
     */
    public static void main(String[] args) {
        String regexp = "(" + args[0] + ")";
        String txt = args[1];
        NFA nfa = new NFA(regexp);
        StdOut.println(nfa.recognizes(txt));
    }
}

File: Kotlin/src/me/metatom/kotlin/string/NFA.kt
package me.metatom.kotlin.string

import edu.princeton.cs.algs4.Bag
import edu.princeton.cs.algs4.Digraph
import edu.princeton.cs.algs4.DirectedDFS
import edu.princeton.cs.algs4.Stack

/**
 * 正则表达式的模式匹配
 *
 * @author igaozp
 * @since 2017-07-28
 * @version 1.0
 */
class NFA(regexp: String) {
    /**
     * 匹配转换
     */
    private var re: Array<Char?>? = null
    /**
     * epsilon 转换
     */
    private var G: Digraph? = null
    /**
     * 状态数量
     */
    private var M: Int = 0

    /**
     * 构造方法，构建正则表达式的 NFA
     */
    init {
        val ops = Stack<Int>()
        val re = regexp.toCharArray()
        M = re.size
        G = Digraph(M + 1)
        for (i in 0 until M) {
            var  lp = i
            if (re[i] == '(' || re[i] == '|') {
                ops.push(i)
            } else if (re[i] == ')') {
                val or = ops.pop()
                if (re[or] == '|') {
                    lp = ops.pop()
                    G!!.addEdge(lp, or + 1)
                    G!!.addEdge(or, i)
                } else {
                    lp = or
                }
            }

            if (i < M - 1 && re[i + 1] == '*') {
                G!!.addEdge(lp, i + 1)
                G!!.addEdge(i + 1, lp)
            }

            if (re[i] == '(' || re[i] == '*' || re[i] == ')') {
                G!!.addEdge(i, i + 1)
            }
        }
    }

    /**
     * NFA 能否检查文本
     *
     * @param txt 检查的文本
     * @return `true` 能够检查
     *         `false` 不能检查
     */
    fun recognizes(txt: String): Boolean {
        var pc = Bag<Int>()
        var dfs = DirectedDFS(G, 0)

        (0 .. G!!.V()).forEach { v ->
            if (dfs.marked(v)) {
                pc.add(v)
            }
        }

        for (i in 0 until txt.length) {
            val match = Bag<Int>()
            pc.forEach { v ->
                if (v < M) {
                    if (re!![v] == txt[i] || re!![v] == '.') {
                        match.add(v + 1)
                    }
                }
            }

            pc = Bag()
            dfs = DirectedDFS(G, match)

            (0 until G!!.V()).forEach { v ->
                if (dfs.marked(v)) {
                    pc.add(v)
                }
            }
        }

        pc.forEach { v ->
            if (v == M) {
                return true
            }
        }

        return false
    }
}